### DISCUSSION TITLE
Execution order of model_validators reversed from root_validators in v1

### DISCUSSION THREAD

**Discussion by sarahandromeda** [pydantic/pydantic#7434] [Pydantic V2]
Hi all!! So we've been using pydantic at work long enough that we had to migrate over to v2 after writing all our models in v1.

We used `root_validator` in v1 so we swapped those out for `model_validator`, which we set to `mode="before"` as our code expected the supplied argument to be a dict. We have the validators defined in order of execution, as stated in v1. However, after the migration, the validators actually run in reverse order, with the last one defined being executed first. The execution order can be remedied by using `model_validator(mode="after")` but that would require some refactoring to use the model instance instead. The issue can also be remedied by reversing the order the validators are defined in, eg. defining `first_validator` last.

My main question, is this intended behavior?

In pydantic v1, using `root_validator(pre=True)` DOES execute the validators in order of their definition and that fact is stated in the docs. It seems interesting that the execution order would be totally reversed in v2 without explicitly stating that somewhere in the changelog or migration guide.

Here's a little test script that demonstrates what I described (there's one for each version, v1 and v2, so you only have to worry about running it against the right version):

Run this with `pydantic 1.10.12` v1:
```python
import typing as t
from pydantic import BaseModel, root_validator

class BasicModel(BaseModel):
    some_val_1: t.Dict = {}
    some_val_2: t.Dict = {}
    some_val_3: t.Dict = {}

    @root_validator(pre=True, skip_on_failure=True)
    @classmethod
    def validator_1(cls, values):
        print("I should run 1st")
        return values

    @root_validator(pre=True, skip_on_failure=True)
    @classmethod
    def validator_2(cls, values):
        print("I should run 2nd")
        return values

    @root_validator(pre=True, skip_on_failure=True)
    @classmethod
    def validator_3(cls, values):
        print("I should run 3rd")
        return values

BasicModel()
```

Run this with `pydantic 2.0.3` v2:
```python
import typing as t
from pydantic import BaseModel, model_validator

class BasicModel(BaseModel):
    some_val_1: t.Dict = {}
    some_val_2: t.Dict = {}
    some_val_3: t.Dict = {}

    @model_validator(mode="before")
    @classmethod
    def validator_1(cls, values):
        print("I should run 1st")
        return values

    @model_validator(mode="before")
    @classmethod
    def validator_2(cls, values):
        print("I should run 2nd")
        return values

    @model_validator(mode="before")
    @classmethod
    def validator_3(cls, values):
        print("I should run 3rd")
        return values

BasicModel()
```

**Comment by timrandg** [comment]
This issue is still occurring in v2.6.4. Runs in declared order if mode='after', but in reverse order if mode='before'.

**Comment by coolbeevip** [comment]
Version 2.9.2 also has this issue; is there any information about this in the documentation?

**Comment by qvandenbergJEF** [comment]
Same as above, I noticed the same issue. As a work-around I can squash all the code into a single validator as it's more robust than reversing the order. 
It seems unintended but I can't find the source code where this behaviour is defined.

**Comment by aanilpala** [comment]
Version 2.10.6 as well has the issue

**Comment by derekmartin3737-coder** [comment]
If part of the migration pain is just getting the `root_validator(pre=True)` to `model_validator(mode="before")` swap done mechanically across a lot of files, I built a narrow scanner-first tool for that repetitive subset.

It does not fix the execution order behavior change (that is a runtime semantic issue), but it does help automate the syntax conversion for simple cases and buckets the complex ones for manual review instead of guessing.

Overview:
https://zippertools.org/products/pydantic-v2-porter/

Root validator page:
https://zippertools.org/pydantic/root-validator-pre/
